sockeye/training.py

```diff
@@ -135,17 +185,73 @@ def sym_gen(seq_lens):
                                                            source_embed_length,
                                                            source_embed_seq_len)
 
-            # decoder
-            # target_decoded: (batch-size, target_len, decoder_depth)
-            target_decoded = self.decoder.decode_sequence(source_encoded, source_encoded_length, source_encoded_seq_len,
+            # target embedding
+            (target_embed,
+             target_embed_length,
+             target_embed_seq_len) = self.embedding_target.encode(target, target_length, target_seq_len)
+
+            if self.config.scheduled_sampling_type is not None:
+                # use scheduled sampling: decode iteratively per-token
+
+                # logits_all: target_seq_len * (batch_size, target_vocab_size)
+                logits_all = []
+
+                # threshold over which we draw samples instead of using the gold target previous word history
+                threshold = self._get_sampling_threshold()
+
+                self.decoder.reset()
+                source_encoded_batch_major = mx.sym.swapaxes(source_encoded, dim1=0, dim2=1,
+                                                             name='source_encoded_batch_major')
+                states = self.decoder.init_states(source_encoded_batch_major,
+                                                               source_encoded_length,
+                                                               source_encoded_seq_len)
+
+                # slice target words
+                # target: target_seq_len * (batch_size)
+                target_sliced = mx.sym.split(data=target, num_outputs=target_seq_len, axis=1, squeeze_axis=True)
+
+                for seq_idx in range(target_embed_seq_len):
+                    if seq_idx == 0:
+                        target_prev = target_sliced[seq_idx]
+
+                    else:
+                        rand_val = mx.sym.random_uniform(low=0, high=1.0, shape=1)[0]
+                        target_prev = mx.sym.where(rand_val < threshold,
+                                                     mx.sym.cast(target_sliced[seq_idx], dtype='int32'), sampled_word)
```

Comment:
its a bit unclear how sampled_word is initialized due to the special case if seq_idx==0.

Reply:
It could also use some shape information in comments

Reply:
It's a bit confusing at first to see sampled_word used before it's initialized (after realizing that it's a loop it becomes clear, but could be something to rewrite for readability).

If I understood it right, target_prev is either the current, **seq_idx**'th, target word or the sampled word from the (**seq_idx-1**)'th distribution, since sampling happened in the previous iteration. Is it intended, shouldn't we chose between a target word and a sample from the distribution from the same position?